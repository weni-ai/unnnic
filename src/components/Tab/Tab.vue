<template>
  <div :class="['tab', `size-${size}`]">
    <header class="tab-header">
      <ul class="tab-content">
        <li
          class="tab-head"
          v-for="tab in tabs"
          :key="tab"
          v-bind:class="{
            'tab-head--active': localValue === tab,
          }"
          @click="change(tab)"
        >
          <slot :name="tabHeadSlotName(tab)">{{ tab }} </slot>
        </li>
      </ul>
    </header>
    <main class="tab-body">
      <div class="tab-panel"><slot :name="tabPanelSlotName"> </slot></div>
    </main>
  </div>
</template>

<script>
export default {
  model: {
    prop: 'activeTab',
    event: 'change',
  },
  props: {
    size: {
      type: String,
      default: 'md',
      validator: (size) => ['md', 'sm'].includes(size),
    },
    activeTab: {
      type: String,
    },
    initialTab: {
      type: String,
      default: null,
    },
    tabs: {
      type: Array,
      default: null,
    },
  },
  data() {
    return {
      localValue: '',
    };
  },
  created() {
    const value = this.initialTab || this.activeTab || (this.tabs && this.tabs[0]);

    this.change(value);
  },
  computed: {
    tabPanelSlotName() {
      return `tab-panel-${this.localValue}`;
    },
  },
  watch: {
    activeTab() {
      this.localValue = this.activeTab;
    },
  },
  methods: {
    tabHeadSlotName(tabName) {
      return `tab-head-${tabName}`;
    },
    change(value) {
      this.localValue = value;
      this.$emit('change', this.localValue);
    },
  },
};
</script>

<style lang="scss" scoped>
@import "../../assets/scss/unnnic.scss";

.tab-header {
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
  color: $unnnic-color-neutral-cleanest;
  font-family: $unnnic-font-family-secondary;
  font-weight: $unnnic-font-weight-bold;
  font-size: $unnnic-font-size-body-lg;
  line-height: ($unnnic-font-size-body-lg + $unnnic-line-height-medium);
  padding-bottom: $unnnic-inset-sm;
  margin-bottom: $unnnic-inset-sm;
  border-bottom: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
}

.tab-content {
  display: flex;
  margin: 0;
  padding: 0;
  list-style: none;
}

.tab-head {
  cursor: pointer;
  margin-right: $unnnic-inset-md;
}

.tab-head--active {
  font-family: $unnnic-font-family-secondary;
  font-weight: $unnnic-font-weight-bold;
  color: $unnnic-color-neutral-darkest;
  font-size: $unnnic-font-size-body-lg;
  line-height: ($unnnic-font-size-body-lg + $unnnic-line-height-medium);
  transition: 0.4s;
}

.tab.size-sm {
  .tab-header {
    padding-bottom: $unnnic-spacing-stack-xs;
    margin-bottom: $unnnic-spacing-stack-xs;

    .tab-head, .tab-head--active {
      font-size: $unnnic-font-size-body-md;
      line-height: ($unnnic-font-size-body-md + $unnnic-line-height-md);
    }
  }
}
</style>
