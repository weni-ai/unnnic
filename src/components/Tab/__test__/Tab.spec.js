import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, afterEach, test, vi } from 'vitest';
import Tab from '@/components/Tab/Tab.vue';
import UnnnicIcon from '@/components/Icon.vue';
import UnnnicToolTip from '@/components/ToolTip/ToolTip.vue';

describe('Tab.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(Tab, {
      props: {
        size: 'md',
        activeTab: 'tab1',
        initialTab: 'tab2',
        tabs: ['tab1', 'tab2', 'tab3'],
      },
      slots: {
        'tab-head-tab1-tooltip': 'Tooltip text',
      },
      global: {
        components: {
          UnnnicIcon,
          UnnnicToolTip,
        },
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('renders correct number of tabs', () => {
    const tabs = wrapper.findAll('.tab-head');
    expect(tabs.length).toBe(3);
  });

  test('applies correct active class based on activeTab', async () => {
    const activeTab = wrapper.find('.tab-head--active');
    expect(activeTab.text()).toContain('tab2');

    await wrapper.setProps({ activeTab: 'tab3' });
    const newActiveTab = wrapper.find('.tab-head--active');
    expect(newActiveTab.text()).toContain('tab3');
  });

  test('renders UnnnicToolTip and UnnnicIcon when tooltip is provided', () => {
    const tooltips = wrapper.findAllComponents(UnnnicToolTip);
    const icons = wrapper.findAllComponents(UnnnicIcon);

    expect(tooltips.length).toBeGreaterThan(0);
    expect(icons.length).toBeGreaterThan(0);
  });

  test('does not render UnnnicToolTip when no tooltip is provided', async () => {
    await wrapper.setProps({ tabs: ['tab2', 'tab3'] });

    const tooltips = wrapper.findAllComponents(UnnnicToolTip);
    expect(tooltips.length).toBe(0);
  });

  test('correctly computes tabPanelSlotName', () => {
    expect(wrapper.vm.tabPanelSlotName).toBe('tab-panel-tab2');
    wrapper.setData({ localValue: 'tab3' });
    expect(wrapper.vm.tabPanelSlotName).toBe('tab-panel-tab3');
  });

  test('correctly computes tabHeadSlotName', () => {
    expect(wrapper.vm.tabHeadSlotName('tab1')).toBe('tab-head-tab1');
  });

  test('correctly computes getHeadTooltip', () => {
    const tooltip = wrapper.vm.getHeadTooltip('tab2');
    expect(tooltip).toBe('');
  });

  test('changes tab when clicking on tab head', async () => {
    const tabHeads = wrapper.findAll('.tab-head');
    await tabHeads[1].trigger('click');

    expect(wrapper.emitted('change')).toBeTruthy();
    expect(wrapper.emitted('change')[0]).toEqual(['tab2']);
  });

  test('initializes with correct tab based on props', () => {
    expect(wrapper.vm.localValue).toBe('tab2');
  });

  test('updates localValue when activeTab prop changes', async () => {
    await wrapper.setProps({ activeTab: 'tab3' });
    expect(wrapper.vm.localValue).toBe('tab3');
  });

  test('sets the correct initial tab value on creation', () => {
    wrapper = mount(Tab, {
      props: {
        initialTab: 'tab3',
        activeTab: '',
        tabs: ['tab1', 'tab2', 'tab3'],
      },
    });

    expect(wrapper.vm.localValue).toBe('tab3');

    wrapper = mount(Tab, {
      props: {
        initialTab: '',
        activeTab: 'tab2',
        tabs: ['tab1', 'tab2', 'tab3'],
      },
    });

    expect(wrapper.vm.localValue).toBe('tab2');

    wrapper = mount(Tab, {
      props: {
        initialTab: '',
        activeTab: '',
        tabs: ['tab1', 'tab2', 'tab3'],
      },
    });

    expect(wrapper.vm.localValue).toBe('tab1');
  });

  test('computes tabHeadTooltipSlotName correctly', () => {
    expect(wrapper.vm.tabHeadTooltipSlotName('tab1')).toBe(
      'tab-head-tab1-tooltip',
    );
    expect(wrapper.vm.tabHeadTooltipSlotName(undefined)).toBe('');
  });

  test('renders UnnnicIcon with correct size based on prop size', async () => {
    const icon = wrapper.findComponent(UnnnicIcon);
    expect(icon.props('size')).toBe('sm');

    await wrapper.setProps({ size: 'sm' });
    expect(icon.props('size')).toBe('xs');

    await wrapper.setProps({ size: 'lg' });
    expect(icon.props('size')).toBe('sm');
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
