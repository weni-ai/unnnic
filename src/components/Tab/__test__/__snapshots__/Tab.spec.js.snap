// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tab.vue > matches the snapshot 1`] = `
"<div data-v-b4e39fac="" class="tab size-md">
  <header data-v-b4e39fac="" class="tab-header">
    <ul data-v-b4e39fac="" class="tab-content">
      <li data-v-b4e39fac="" class="tab-head">tab1<div data-v-bf0cf546="" data-v-b4e39fac="" class="unnnic-tooltip"><span data-v-26446d8e="" data-v-b4e39fac="" class="material-symbols-rounded unnnic-icon-scheme--neutral-cleanest unnnic-icon-size--sm material-symbols-rounded--filled" data-testid="material-icon" translate="no">info</span><span data-v-bf0cf546="" class="unnnic-tooltip-label unnnic-tooltip-label-bottom" data-testid="tooltip-label" style="left: 0px; top: 8px;">Tooltip text <br data-v-bf0cf546=""><!--v-if--></span></div>
      </li>
      <li data-v-b4e39fac="" class="tab-head tab-head--active">tab2
        <!--v-if-->
      </li>
      <li data-v-b4e39fac="" class="tab-head">tab3
        <!--v-if-->
      </li>
    </ul>
  </header>
  <main data-v-b4e39fac="" class="tab-body">
    <div data-v-b4e39fac="" class="tab-panel"></div>
  </main>
</div>"
`;
